### PR TITLE
Enable BFT meta data persistency test in the simpleTest

### LINKED ISSUE #152: Use DebugPersistentStorage instead of FileStorage for SimpleKVBC TesterReplica
**Describe the bug**
In order to fix the skvbc tests, I added file persistence for bft metadata to the SimpleKVBC TesterReplica. However, the application state itself is not persisted. This is confusing, and can potentially lead to weird errors. The file persistence should be removed and instead, the in memory, debug persistence should be used.

Thanks to @guyg8 for catching this silly mistake on my part.

**To Reproduce**
N/A

**Expected behavior**
We should use Debug persistence flag.

### PR DESCRIPTION
**Pull request by salieri11**
This PR introduces Simple Test refactoring which allows to run persistency tests in memory.
In essence:
* Test replica can be run in 3 persistency modes: no persistency, file based persistency and in-memory
* Another parameter for the replica is ReplicaBehavior, which defines the test case. ReplicaBehavior can model various replica's shutdowns, with or without view change, one time or periodical
* The test client runs as before, just trying to complete all it's transactions

Additionally, new unit test, persistency_test, performs sanity test without persistency being enable and one test where primary restarts and view change occurs.

Bug fixed:
* Incorrect behavior of Stop() method in the TLS comm module

Notes:
* The simpleKVBC test should be run with in-memory persistency hardcoded. Allowing 
different persistency modes for this test is out of scope for this PR.
* The persistency_test unit test may take about 2.5 minutes to complete 

Testing:
* All unit tests pass
* Running manually simpleTest with all persistency modes and replica behaviours pass

Closes #152

### CODE REVIEW

File: bftengine/tests/simpleTest/persistency_test.cpp [hunk 1]
```diff
@@ -0,0 +1,99 @@
+//
+// Created by Igor Golikov on 2019-07-01.
```
Comment (andrewjstone) [inline bftengine/tests/simpleTest/persistency_test.cpp]: This is information that can be approximately be discovered via git. Feel free to leave it in though. 

We should also have a VMware copyright header in this file.

File: bftengine/src/bftengine/BFTEngine.cpp [hunk 2]
```diff
@@ -119,13 +131,13 @@ Replica *Replica::createNewReplica(ReplicaConfig *replicaConfig,
   uint16_t numOfObjects = 0;
   bool isNewStorage = true;
 
-  // metadataStorage is essential for a non-testing mode.
-  if (!debugPersistentStorageEnabled) Assert(metadataStorage != nullptr);
+  if (replicaConfig->usePedanticPersistencyChecks) {
```
[the remark this answers is not in the document]
Comment (salieri11) [reply bftengine/src/bftengine/BFTEngine.cpp]: Well maybe the name usePedanticChecks is incorrect. It should be debugPersistentStorageEnabled... the same one that was replaced (the old one lived in the SysConsts file). Then the assert makes sense.

File: bftengine/tests/simpleTest/README.md [hunk 3]
```diff
@@ -120,6 +120,10 @@ Note: if you run the command as described above and experience Python related
 
     python3 build/bftengine/tests/simpleTest/scripts/simpleTest.py -bft n=4,r=4,f=1,c=0,cl=1
 
+### BFT Metadata with the simple test
+BFT metadata allows replica to recover from crash if the metadata has been written to the disc when replica is running.
```
[the remark this answers is not in the document]
Comment (salieri11) [reply bftengine/tests/simpleTest/README.md]: thanks!

File: bftengine/tests/CMakeLists.txt [hunk 4]
```diff
@@ -1,7 +1,8 @@
 set(CONFIG_FOLDER_PATH_VALUE "${CMAKE_CURRENT_SOURCE_DIR}/config")
-add_subdirectory(simpleStorage)
+add_subdirectory(config)
 add_subdirectory(simpleTest)
 add_subdirectory(simpleKVBC)
 add_subdirectory(simpleKVBCTests)
+add_subdirectory(simpleStorage)
```
Comment (yuliasherman) [inline bftengine/tests/CMakeLists.txt]: Did you intentionally move simpleStorage down? Tests depend on it, thus I moved it to the beginning in the past.

Comment (salieri11) [reply bftengine/tests/CMakeLists.txt]: reverted

File: bftengine/tests/simpleTest/CMakeLists.txt [hunk 5]
```diff
@@ -30,20 +33,48 @@ add_executable(configTest
...
 if(USE_LOG4CPP)
     target_compile_definitions(simpleTest_server PUBLIC USE_LOG4CPP=1)
     target_compile_definitions(simpleTest_client PUBLIC USE_LOG4CPP=1)
     target_compile_definitions(configTest PUBLIC USE_LOG4CPP=1)
 endif()
 
 if(${USE_COMM_PLAIN_TCP})
-    target_compile_definitions(simpleTest_client PUBLIC USE_COMM_PLAIN_TCP)
-    target_compile_definitions(simpleTest_server PUBLIC USE_COMM_PLAIN_TCP)
+    target_compile_definitions(simnpleTest_client PUBLIC USE_COMM_PLAIN_TCP)
```
[the remark this answers is not in the document]
Comment (salieri11) [reply bftengine/tests/simpleTest/CMakeLists.txt]: fixed

File: bftengine/src/bftengine/BFTEngine.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (salieri11) [reply bftengine/src/bftengine/BFTEngine.cpp]: renamed

File: bftengine/tests/simpleTest/CMakeLists.txt (hunk 5 above)
[the remark this answers is not in the document]
Comment (salieri11) [reply bftengine/tests/simpleTest/CMakeLists.txt]: fixed

**Pull request merged**